### PR TITLE
Create index for sorted activity

### PR DESCRIPTION
**Pull request by benthecarman**
Now keeps a sorted index of our activity at all time. This should make our activity much faster with only a slight overhead increase.

### CODE REVIEW

**Comment by benthecarman** [comment]
Changed to `limit`/`offset` instead of `start`/`until` as we talked irl

File: mutiny-core/src/lib.rs [hunk 1]
```diff
@@ -1369,39 +1429,94 @@ impl<S: MutinyStorage> MutinyWallet<S> {
...
+                        let user_channel_id_str = item
+                            .key
+                            .trim_start_matches(CHANNEL_CLOSURE_PREFIX)
+                            .splitn(2, '_') // Channel closures have `_{node_id}` at the end
+                            .collect::<Vec<&str>>()[0];
+                        let user_channel_id: [u8; 16] = FromHex::from_hex(user_channel_id_str)?;
+                        closure.user_channel_id = Some(user_channel_id);
+                    }
+                    activities.push(ActivityItem::ChannelClosed(closure));
+                }
```
Comment (TonyGiorgio) [inline mutiny-core/src/lib.rs:1510]: We have `list_channel_closures` which is now no longer used or accessed, we should move the logic here into a method that lists channel closes which can be reusable, and remove the old one.

File: mutiny-core/src/lib.rs [hunk 2]
```diff
@@ -1369,39 +1429,94 @@ impl<S: MutinyStorage> MutinyWallet<S> {
...
+            match (limit, offset) {
+                (None, None) => vec,
+                (Some(limit), Some(offset)) => {
+                    let start = offset;
+                    let end = start.saturating_add(limit).min(vec.len());
+                    vec[start..end].to_vec()
+                }
+                (None, Some(offset)) => vec[offset..].to_vec(),
+                (Some(limit), None) => vec[..limit].to_vec(),
+            }
```
Comment (TonyGiorgio) [inline mutiny-core/src/lib.rs:1478]:
```suggestion
            let (start, end) = match (limit, offset) {
                (None, None) => (0, vec.len()),
                (Some(limit), Some(offset)) => {
                    let start = offset.min(vec.len());
                    let end = start.saturating_add(limit).min(vec.len());
                    (start, end)
                }
                (None, Some(offset)) => (offset.min(vec.len()), vec.len()),
                (Some(limit), None) => (0, limit.min(vec.len())),
            };

            vec[start..end].to_vec()
```

I think you can make this easier to read and safer by limiting the scope of what happens in the match. Just make it return the start and end and only need to call `to_vec()` in one place.

**Review by TonyGiorgio — CHANGES_REQUESTED** [review]
Screenshot 2024-02-02 at 11 43 42 AM

An unknown 0 sat on chain payment appeared during some channel closes.

Screenshot 2024-02-02 at 11 47 15 AM

Multiple force close items are showing up when I only had one. 

Further, I cannot test this long enough without getting OOM issues.

**Comment by TonyGiorgio** [comment]
Screenshot 2024-02-02 at 11 50 40 AM

This lightning payment is showing up between the older on chain force closes that are duplicated

**Comment by benthecarman** [comment]
Fixed all the duplication errors which also seemed to fixed the crashing for me too, curious if it helps for you as well.

**Comment by TonyGiorgio** [comment]
What if we did this and defaulted the activity list to the last ~30 items? 

My phone is starting to get very slow. And having the new social home screen basically being a "view all"  for the transaction list, I would rather it not be that much.

**Comment by benthecarman** [comment]
Makes sense to me, can rebase

**Review by TonyGiorgio — APPROVED** [review]
Code looks good. If you just want to test it to make sure it works with current mutiny master, that would be great.

**Comment by benthecarman** [comment]
Found a potential index out of bounds error when testing, fixed it and added tests. Otherwise everything else felt good

**Review by TonyGiorgio — APPROVED** [review]

**Comment by benthecarman** [comment]
sorry last push, chatgpt told me another check to add just in case:

```
// handle start > end
if start > end {
	return vec![];
}
```

**Pull request merged**